lib/mastodon/accounts_cli.rb

```diff
@@ -509,6 +509,47 @@ def approve(username = nil)
       end
     end
 
+    option :concurrency, type: :numeric, default: 5, aliases: [:c]
+    option :dry_run, type: :boolean
+    desc 'prune', 'Prune remote accounts that never interacted with local users'
+    long_desc <<-LONG_DESC
+      Prune remote account that
+      - follows no local accounts
+      - is not followed by any local accounts
+      - has no statuses on local
+      - has not been mentioned
+      - has not been favourited local posts
+      - not muted/blocked by us
+    LONG_DESC
+    def prune
+      dry_run = options[:dry_run] ? ' (dry run)' : ''
+
+      query = Account.remote.where.not(actor_type: %i(Application Service))
+      query = query.where.missing(:mentions)
+      query = query.where.missing(:favourites)
+      query = query.where.missing(:statuses)
+      query = query.joins('left outer join follows as a on accounts.id = a.account_id')
+      query = query.joins('left outer join follows as b on accounts.id = b.target_account_id')
```

Comment:
The following might have better performances:
```ruby
  query = query.where('NOT EXISTS (SELECT 1 FROM mentions WHERE account_id = accounts.id)')
  query = query.where('NOT EXISTS (SELECT 1 FROM favourites WHERE account_id = accounts.id)')
  query = query.where('NOT EXISTS (SELECT 1 FROM statuses WHERE account_id = accounts.id)')
  query = query.where('NOT EXISTS (SELECT 1 FROM follows WHERE account_id = accounts.id OR target_account_id = accounts.id)')
```

Also I think you should check for `follow_requests`. Note that `follow_requests` does not have an index directly on `target_account_id`, so this may prove expensive to check.